src/System.Private.ServiceModel/tests/Common/Scenarios/Endpoints.cs

```diff
@@ -1,49 +1,183 @@
+                    serviceProxy = factory.CreateChannel();
+                    clientCert = new X509Certificate2(serviceProxy.GetClientCert(false), "test", X509KeyStorageFlags.PersistKeySet);
+                    if (clientCert == null)
+                    {
+                        //throw
+                        throw new Exception("Failed to obtain client cert from the server");
+                    }
+
+                    BridgeClientCertificateManager.AddToStoreIfNeeded(StoreName.My, StoreLocation.CurrentUser, clientCert);
+                    BridgeClientCertificateManager.LocalCertThumbprint = clientCert.Thumbprint;
+                }
+                finally
+                {
+                    ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)serviceProxy, factory);
+                }
+            }
+        }
+    }
+
+    private static string GetServiceHostName()
+    {
+        return TestProperties.GetProperty(TestProperties.BridgeHost_PropertyName);
+    }
+
+    private static bool IISHosted
+    {
+        get
+        {
+            if (TestProperties.GetProperty(TestProperties.ServiceUri_PropertyName).ToLower()==("localhost"))
+            {
+                return false;
+            };
+
+            return true;
+        }
+    }
+
+    private static Uri BuildBaseUri(string protocol)
+    {
+        var builder = new UriBuilder();
+        builder.Host = TestProperties.GetProperty(TestProperties.ServiceUri_PropertyName);
+        builder.Scheme = protocol;
+
+        if (!IISHosted)
+        {
+            switch (protocol)
+            {
+                case "http":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeHttpPort_PropertyName));
+                    break;
+                case "ws":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeWebSocketPort_PropertyName));
+                    builder.Scheme = "http";
+                    break;
+                case "https":
+                     builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeHttpsPort_PropertyName));
+                    break;
+                case "wss":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeSecureWebSocketPort_PropertyName));
+                    builder.Scheme = "https";
+                    break;
+                case "net.tcp":
+                    builder.Port = int.Parse(TestProperties.GetProperty(TestProperties.BridgeTcpPort_PropertyName));
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        return builder.Uri;
+    }
+    public static string GetEndpointAdddress(string endpoint, string protocol = "http")
+    {
+        return string.Format(@"{0}/{1}", BuildBaseUri(protocol), endpoint);
+    }
+
+    public static string ServiceUtil_Address
+    {
```

Comment:
Not blocking: we need a better name for the service interface, class and endpoint name.  Maybe CertificateService_Address.